chimney/src/main/scala/io/scalaland/chimney/internal/macros/TransformerMacros.scala

```diff
@@ -81,58 +83,67 @@ trait TransformerMacros extends TransformerConfiguration {
   def expandTransformerTree(
       srcPrefixTree: Tree,
       config: TransformerConfig
-  )(From: Type, To: Type): Either[Seq[DerivationError], Tree] = {
+  )(
+      From: Type,
+      To: Type,
+      F: Type
+  ): Either[Seq[DerivationError], Tree] = {
 
     val localInstance: Option[Tree] =
       if (config.definitionScope.contains((From, To))) {
         None
       } else {
-        findLocalImplicitTransformer(From, To)
+        findLocalImplicitTransformerF(F, From, To)
       }
 
+    val ctx = q"_root_.io.scalaland.chimney.TransformationContext[$F]"
+
     localInstance
       .map { localImplicitTree =>
         Right(q"$localImplicitTree.transform($srcPrefixTree)")
       }
       .getOrElse {
         if (isSubtype(From, To)) {
-          Right(srcPrefixTree)
+          Right(q"$ctx.pure[$To]($srcPrefixTree)")
         } else if (fromValueClassToType(From, To)) {
-          expandValueClassToType(srcPrefixTree)(From, To)
+          expandValueClassToType(srcPrefixTree, ctx)(From, To)
         } else if (fromTypeToValueClass(From, To)) {
-          expandTypeToValueClass(srcPrefixTree)(From, To)
+          expandTypeToValueClass(srcPrefixTree, ctx)(From, To)
         } else if (bothOptions(From, To)) {
-          expandOptions(srcPrefixTree, config)(From, To)
+          expandOptions(srcPrefixTree, config, ctx)(From, To, F)
         } else if (isOption(To)) {
-          expandTargetWrappedInOption(srcPrefixTree, config)(From, To)
-        } else if (config.enableUnsafeOption && isOption(From)) {
-          expandSourceWrappedInOption(srcPrefixTree, config)(From, To)
-        } else if (bothEithers(From, To)) {
+          expandTargetWrappedInOption(srcPrefixTree, config, ctx)(From, To, F)
+        } else if (isOption(From)) {
+          expandSourceWrappedInOption(srcPrefixTree, config, ctx)(From, To, F)
+        } else if (bothEithers(From, To) && isId(F)) {
           expandEithers(srcPrefixTree, config)(From, To)
-        } else if (bothMaps(From, To)) {
+        } else if (bothMaps(From, To) && isId(F)) {
           expandMaps(srcPrefixTree, config)(From, To)
-        } else if (bothOfTraversableOrArray(From, To)) {
+        } else if (bothOfTraversableOrArray(From, To) && isId(F)) {
```

Comment:
It should be possible to generate collections (also maps, eithers, etc.) transformers for any `F`, but you would need `traverse` in `TransformerContext`.

Reply:
Yes, I was thinking about this, but it's non-optimal to do this for any iterable.
I want to safe bothOfTraversableOrArray for backward compatibility in id context and require Traverse[G] instance in other contexts.

Reply:
@krzemin, I implemented Traverse for cats.Traverse collections and fixed all old test cases. The only place where compatibility was broken is TransformationException instead of NoSuchElementException in enableUnsafeOptions. Also I think it's should be support of maps in F context